File: Networking assignment arunava/C/hamming.c
#include<stdio.h>
#include<conio.h>
#include<string.h>
#include<math.h>
void calc_redundancy(char *data_new2,int m,int r);
char red[10];
int arr[30];
void main()
{
	char data[50],dred[80],dnew[80],derror[80],new_red_bit[15],temp[80];
	int m,r,r1,i,j=0,k=0,error,pos=0;
	printf("Enter the message: ");
	gets(data);
	m=strlen(data);
	for(r1=1;r1<m;r1++)
	{
		if (m+r1+1<=pow(2,r1))
		{
			r=r1;
			break;
		}
	}
	printf("Number of redundant bit = %d",r);
	for(i=0;i<r;i++)
		arr[j++]=pow(2,i);
	j=r-1;
	for(i=r+m;i>0;i--)
	{
		if(i==arr[j])
			{
				dred[r+m-i]='x';
				j--;
			}			
		else
			dred[r+m-i]=data[k++];
	}
	dred[r+m]='\0';
	printf("\n\nData with redundant : %s",dred);
	strcpy(dnew,dred);
	strrev(dnew);
	calc_redundancy(dnew,m,r);
	printf("\nRedundant bits are----> %s",red);
	printf("\n\nData with redundant : %s",dnew);
	printf("\nEnter the error position from MSB(If no error enter 0) = ");
	scanf("%d",&error);
	strcpy(derror,dnew);
	strrev(derror);
	if(error!=0)
	{
		if(derror[error-1]=='0')
			derror[error-1]='1';
		else
			derror[error-1]='0';
	}
	strcpy(temp,derror);
	printf("\nReceived data with error: %s",strrev(temp));
	calc_redundancy(derror,m,r);
	k=0;
	strcpy(new_red_bit,red);
	for(i=0;i<r;i++)
	{
		pos+=pow(2,i)*(new_red_bit[r-i-1]-'0');
	}
	printf("\nNew Redundant bits are----> %s\n",new_red_bit);
	printf("\nError position = %d",pos);
	if(derror[pos-1]=='0')
		derror[pos-1]='1';
	else
		derror[pos-1]='0';
	printf("\nReceived data after correction: %s",strrev(derror));
	getch();
}
void calc_redundancy(char *data_new2,int m,int r)
{
	int i,p,q,j,k,count_1,b,h=0;
	char eqn[80];
	for(b=0;b<r;b++)
	{
		p=b+1;
		q=pow(2,p);	
		eqn[0]='\0';
		k=0;
		for(i=arr[b]-1;i<m+r;i+=q)
		{
			if(i+arr[b]<=m+r)
			{
				for(j=i;j<i+arr[b];j++)
				eqn[k++]=data_new2[j];
			}
			else
			{
				for(j=i;j<m+r;j++)
				eqn[k++]=data_new2[j];
			}
		}
		eqn[k]='\0';
		count_1=0;
		for(i=0;i<strlen(eqn);i++)
		{
			if(eqn[i]=='1')
			count_1++;
		}
		if (count_1%2==0)
		{
			data_new2[(int)(q/2)-1]='0';
			red[h++]='0';
		}	
		else
		{
			data_new2[(int)(q/2)-1]='1';
			red[h++]='1';
		}
		red[h]='\0';
	}
	strrev(data_new2);
	strrev(red);
}

File: Networking assignment arunava/Python/Hamming.py
from math import log
def calc_reduntancy(data_new2,pow_2,m,r):
	redundant={}.fromkeys(pow_2,'x')
	for item in pow_2:
		p=int(log(item,2))+1
		q=2**p
		#print(item,q)
		eqn=''
		for i in range(item-1,m+r,q):
			#print(i+1,i+item)
			try:
				eqn+=data_new2[i:i+item]
			except:
				eqn+=data_new2[i:m+r]
		count_1=eqn.count('1')
		#print(eqn[::-1])
		if count_1%2==0:
			redundant[item]='0'
		else:
			redundant[item]='1'
	return redundant
print("Enter the message")
data=input().strip()
m=len(data)
for r1 in range(1,m):
	if m+r1+1<=2**r1:
		r=r1
		break
print('Number of redundant bit = ',r)
data_new=''
pow_2=[]
for i in range(r):
	pow_2.append(2**i)
redundant={}.fromkeys(pow_2,'x')
j=0
for i in range(r+m,0,-1):
	if i in pow_2:
		data_new+='x'
	else:
		data_new+=data[j]
		j+=1
#print(data_new)
#print(redundant)
data_new2=data_new[::-1]
redundant=calc_reduntancy(data_new2,pow_2,m,r)
data_trans=''
j=0
for i in range(r+m,0,-1):
	if i in pow_2:
		data_trans+=redundant[i]
	else:
		data_trans+=data[j]
		j+=1
print("Redundant bits are---->")
for item in pow_2:
	print('c['+str(item)+'] = ',redundant[item])
print('Transmitting data = ',data_trans)
print("Enter the error position (If no error enter 0)")
error=int(input().strip())
data_trans2=data_trans[::-1]
data_error=''+data_trans2
if error!=0:
	if data_trans2[error-1]=='1':
		error_bit='0'
	else:
		error_bit='1'
	data_error=data_trans2[:error-1]+error_bit+data_trans2[error:]
new_redundant=calc_reduntancy(data_error,pow_2,m,r)
new_red_bit=''
for item in pow_2:
	new_red_bit+=new_redundant[item]
new_red_bit=new_red_bit[::-1]
print("New Redundant bits are----> ",new_red_bit)
pos=int(new_red_bit,2)
print("Error position = ",pos)